addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/CORSActiveScanRule.java

```diff
@@ -0,0 +1,162 @@
+
+    @Override
+    public void scan() {
+        HttpMessage baseMsg = getBaseMsg();
+        URI uri = baseMsg.getRequestHeader().getURI();
+        String authority = uri.getEscapedAuthority();
+        String scheme = uri.getScheme();
+        String handyScheme = scheme + "://";
+
+        // Order of likelihood and severity
+        String[] payloads = {
+            handyScheme + RANDOM_NAME + ".com",
+            "null",
+            handyScheme + RANDOM_NAME + "." + authority,
+            handyScheme + authority + "." + RANDOM_NAME + ".com",
+            // URL encoded backtick used to bypass weak Regex matching only alphanumeric chars to
+            // validate the domain: https://www.corben.io/tricky-CORS/
+            handyScheme + authority + "%60" + RANDOM_NAME + ".com",
+            null,
+            handyScheme + authority
+        };
+
+        boolean secScheme = false;
+        if (scheme == "https") {
+            secScheme = true;
+            payloads[5] = "http://" + authority;
+        }
+
+        for (String payload : payloads) {
+            HttpMessage msg = getNewMsg();
+            msg.getRequestHeader().setHeader("Origin", payload);
+            try {
+                sendAndReceive(msg);
+                String acaoKey = HttpHeader.ACCESS_CONTROL_ALLOW_ORIGIN;
+                String acao = msg.getResponseHeader().getHeader(acaoKey);
+                // If there is an ACAO header an alert will be triggered
+                if (acao == null) continue;
+
+                String name = getName();
+                int risk = Alert.RISK_INFO;
+                String evidence = acaoKey + ": " + acao;
+                int wasc = 14;
+                int cwe = 942;
+                String desc = getDescription();
+                boolean vuln = false;
+
+                // Evaluates the risk for this alert
+                if (acao.contains("*")) {
+                    vuln = true;
+                    risk = Alert.RISK_MEDIUM;
+                } else if (acao.contains(RANDOM_NAME)
+                        || acao.contains("null")
+                        || (secScheme && acao.contains("http:"))) {
+                    vuln = true;
+                    // If authenticated AJAX requests are allowed, the risk is higher
+                    String acacKey = "Access-Control-Allow-Credentials";
+                    String acac = msg.getResponseHeader().getHeader(acacKey);
+                    if (acac == null) risk = Alert.RISK_MEDIUM;
+                    else {
+                        risk = Alert.RISK_HIGH;
+                        evidence += "\n" + acacKey + ": " + acac;
+                    }
+                }
+                if (vuln) {
+                    name = getConstantStr("vuln.name");
+                    desc = getConstantStr("vuln.desc");
+                }
+
+                newAlert()
+                        .setName(name)

```

Comment:
The previous name handling could be removed (L80, L105 to L108) and all of that done here:
```suggestion
                        .setName(vuln ? getConstantStr("vuln.name") : getName())

```